potato/activelearning.py

```diff
@@ -0,0 +1,173 @@
+    # classifiers for each). If the user doesn't specify these, we use all of
+    # them.
+    schema_used = []
+    if "active_learning_schema" in al_config:
+        schema_used = al_config["active_learning_schema"]
+
+    cls_kwargs = al_config.get("classifier_kwargs", {})
+    cls_kwargs = al_config.get("classifier_kwargs", {})
+    vectorizer_kwargs = al_config.get("vectorizer_kwargs", {})
+    strategy = al_config["resolution_strategy"]
+
+    # Collect all the current labels
+    instance_to_labels = defaultdict(list)
+    for uas in user_to_annotation_state.values():
+        for iid, annotation in uas.instance_id_to_labeling.items():
+            instance_to_labels[iid].append(annotation)
+
+    # Resolve all the mutiple-annotations to a single one using the provided
+    # strategy to get training data
+    instance_to_label = {}
+    schema_seen = set()
+    for iid, annotations in instance_to_labels.items():
+        resolved = resolve(annotations, strategy)
+
+        # Prune to just the schema we care about
+        if len(schema_used) > 0:
+            resolved = {k: resolved[k] for k in schema_used}
+
+        for s in resolved:
+            schema_seen.add(s)
+        instance_to_label[iid] = resolved
+
+    # Construct a dataframe for easy processing
+    texts = []
+    # We'll train one classifier for each scheme
+    scheme_to_labels = defaultdict(list)
+    text_key = config["item_properties"]["text_key"]
+    for iid, schema_to_label in instance_to_label.items():
+        # get the text
+        text = instance_id_to_data[iid][text_key]
+        texts.append(text)
+        for s in schema_seen:
+            # In some cases where the user has not selected anything but somehow
+            # this is considered annotated, we include some dummy label
+            label = schema_to_label.get(s, "DUMMY:NONE")
+
+            # HACK: this needs to get fixed for multilabel data and possibly
+            # number data
+            label = list(label.keys())[0]
+            scheme_to_labels[s].append(label)
+
+    scheme_to_classifier = {}
+
+    # Train a classifier for each scheme
+    for scheme, labels in scheme_to_labels.items():
+
+        # Sanity check we have more than 1 label
+        label_counts = Counter(labels)
+        if len(label_counts) < 2:
+            logger.warning(
+                (
+                    "In the current data, data labeled with %s has only a"
+                    + "single unique label, which is insufficient for "
+                    + "active learning; skipping..."
+                )
+                % scheme
+            )
+            continue
+
+        # Instantiate the classifier and the tokenizer
+        cls = get_class(al_config["classifier_name"])(**cls_kwargs)
+        vectorizer = get_class(al_config["vectorizer_name"])(**vectorizer_kwargs)
+
+        # Train the classifier
+        clf = Pipeline([("vectorizer", vectorizer), ("classifier", cls)])
```

Comment:
Given that we are training a series of configured classifiers, we should probably background this task in some capacity so it doesn't interfere with startup. This could take a while.